components/director/internal/tenantfetcher/handler.go

```diff
@@ -0,0 +1,214 @@
+		return
+	}
+
+	if err := tx.Commit(); err != nil {
+		log.C(ctx).WithError(err).Errorf("Failed to commit transaction while storing tenant with external ID %s: %v", externalTenantID, err)
+		http.Error(writer, fmt.Sprintf(tenantCreationFailureMsgFmt, externalTenantID), http.StatusInternalServerError)
+		return
+	}
+
+	writer.Header().Set("Content-Type", "text/plain")
+	writer.WriteHeader(http.StatusOK)
+	if _, err := writer.Write([]byte(compassURL)); err != nil {
+		log.C(ctx).WithError(err).Errorf("Failed to write response body for tenant request creation for tenant %s: %v", externalTenantID, err)
+		return
+	}
+}
+
+func (h *handler) DeleteByExternalID(writer http.ResponseWriter, _ *http.Request) {
+	writer.WriteHeader(http.StatusOK)
+}
+
+func (h *handler) tenantInfoFromBody(body []byte) (*model.BusinessTenantMapping, string, error) {
+	tenantId := gjson.GetBytes(body, h.config.TenantProviderTenantIdProperty)
+	if len(tenantId.String()) <= 0 {
+		return nil, "", fmt.Errorf("mandatory tenant ID property %q is missing from request body", h.config.TenantProviderTenantIdProperty)
+	}
+	subdomain := gjson.GetBytes(body, h.config.TenantProviderSubdomainProperty)
+	if len(subdomain.String()) <= 0 {
+		return nil, "", fmt.Errorf("mandatory subdomain property %q is missing from request body", h.config.TenantProviderSubdomainProperty)
+	}
+	customerId := gjson.GetBytes(body, h.config.TenantProviderCustomerIdProperty)
+
+	return &model.BusinessTenantMapping{
+		ID:             h.uidService.Generate(),
+		Name:           tenantId.String(),
+		ExternalTenant: tenantId.String(),
+		Parent:         customerId.String(),
+		Type:           tenantEntity.Account,
+		Provider:       h.config.TenantProvider,
+		Status:         tenantEntity.Active,
+	}, subdomain.String(), nil
+}
+
+func readBody(r *http.Request) ([]byte, error) {
+	ctx := r.Context()
+
+	buf, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	defer func() {
+		if err := r.Body.Close(); err != nil {
+			log.C(ctx).WithError(err).Errorf("Unable to close request body: %v", err)
+		}
+	}()
+
+	return buf, nil
+}
+
+func (h *handler) ensureParentExists(ctx context.Context, parentTenantID, childTenantID string) (string, error) {
+	log.C(ctx).Infof("Ensuring parent tenant with external ID %s for tenant with ID %s exists", parentTenantID, childTenantID)
+	id, err := h.tenantSvc.GetInternalTenant(ctx, parentTenantID)
+	if err != nil && !apperrors.IsNotFoundError(err) && err != sql.ErrNoRows {
+		return "", err
+	}
+	if id != "" {
+		log.C(ctx).Infof("Parent tenant with external ID %s already exists", parentTenantID)
+		return id, nil
+	}
+
+	log.C(ctx).Infof("Creating parent tenant with external ID %s", parentTenantID)
+	tenant := h.customerTenant(parentTenantID)
+	err = h.tenantSvc.CreateManyIfNotExists(ctx, []model.BusinessTenantMapping{tenant})
```

Comment:
I was going to suggest exporting the CreateIfNotExists method from the tenant service and using it instead, but then I saw what it actually does... :D